### PR TITLE
gas golfing

### PR DESCRIPTION
**Pull request by kadenzipfel**

### CODE REVIEW

**Comment by Vectorized** [comment]
Looks interesting! Can you make some gas numbers?

File: contracts/ERC721A.sol [hunk 1]
```diff
@@ -323,6 +322,24 @@ contract ERC721A is IERC721A {
         }
     }
 
+    /**
+     * @dev Checks if val == 0.
+     */
+    function _isZero(uint256 val) private pure returns (bool isZero) {
```
Comment (Vectorized) [inline contracts/ERC721A.sol:328]: Can rename the the return value to `result` for consistency?

File: contracts/ERC721A.sol [hunk 2]
```diff
@@ -163,7 +163,7 @@ contract ERC721A is IERC721A {
      * @dev See {IERC721-balanceOf}.
      */
     function balanceOf(address owner) public view override returns (uint256) {
-        if (_addressToUint256(owner) == 0) revert BalanceQueryForZeroAddress();
+        if (_isZero(owner)) revert BalanceQueryForZeroAddress();
```
Comment (Vectorized) [inline contracts/ERC721A.sol:166]: Can help check if `_isZero(_addressToUint256(owner))` gives the same gas results?

**Comment by ahbanavi** [comment]
> Can you make some gas numbers?  

It doesn't save gas.
Left side is main branch.
image
image
image

<details><summary>Main</summary>
<pre>
·--------------------------------------------------------------|---------------------------|-------------|-----------------------------·
|                     Solc version: 0.8.11                     ·  Optimizer enabled: true  ·  Runs: 800  ·  Block limit: 30000000 gas  │
·······························································|···························|·············|······························
|  Methods
               │
·····································|·························|·············|·············|·············|···············|··············
|  Contract                          ·  Method                 ·  Min        ·  Max        ·  Avg        ·  # calls      ·  usd (avg)  │
·····································|·························|·············|·············|·············|···············|··············
|  ERC721ABurnableMock               ·  approve                ·      30390  ·      50530  ·      40460  ·            2  ·          -  │
·····································|·························|·············|·············|·············|···············|··············
|  ERC721ABurnableMock               ·  burn                   ·      43329  ·     110960  ·      89972  ·           32  ·          -  │
·····································|·························|·············|·············|·············|···············|··············
|  ERC721ABurnableMock               ·  safeMint               ·          -  ·          -  ·     110982  ·           18  ·          -  │
·····································|·························|·············|·············|·············|···············|··············
|  ERC721ABurnableMock               ·  setApprovalForAll      ·      26277  ·      46189  ·      36233  ·            2  ·          -  │
·····································|·························|·············|·············|·············|···············|··············
|  ERC721ABurnableMock               ·  transferFrom           ·      83849  ·      89584  ·      86717  ·            2  ·          -  │
·····································|·························|·············|·············|·············|···············|··············
|  ERC721ABurnableStartTokenIdMock   ·  approve                ·      32527  ·      52667  ·      42597  ·            2  ·          -  │
·····································|·························|·············|·············|·············|···············|··············
|  ERC721ABurnableStartTokenIdMock   ·  burn                   ·      45444  ·     110867  ·      89558  ·           34  ·          -  │
·····································|·························|·············|·············|·············|···············|··············
|  ERC721ABurnableStartTokenIdMock   ·  safeMint               ·          -  ·          -  ·      93860  ·           18  ·          -  │
·····································|·························|·············|·············|·············|···············|··············
|  ERC721ABurnableStartTokenIdMock   ·  setApprovalForAll      ·      26255  ·      46167  ·      36211  ·            2  ·          -  │
·····································|·························|·············|·············|·············|···············|··············
|  ERC721ABurnableStartTokenIdMock   ·  transferFrom           ·      85986  ·      89513  ·      87750  ·            2  ·          -  │
·····································|·························|·············|·············|·············|···············|··············
|  ERC721AGasReporterMock            ·  mintOne                ·      56188  ·      90388  ·      59608  ·           10  ·          -  │
·····································|·························|·············|·············|·············|···············|··············
|  ERC721AGasReporterMock            ·  mintTen                ·      73869  ·     108069  ·      97809  ·           10  ·          -  │
·····································|·························|·············|·············|·············|···············|··············
|  ERC721AGasReporterMock            ·  safeMintOne            ·      58938  ·      93138  ·      76038  ·            2  ·          -  │
·····································|·························|·············|·············|·············|···············|··············
|  ERC721AGasReporterMock            ·  safeMintTen            ·      76620  ·     110820  ·      93720  ·            2  ·          -  │
·····································|·························|·············|·············|·············|···············|··············
|  ERC721AGasReporterMock            ·  transferFrom           ·      44392  ·      86091  ·      54817  ·            4  ·          -  │
·····································|·························|·············|·············|·············|···············|··············
|  ERC721AGasReporterMock            ·  transferTenAsc         ·          -  ·          -  ·     297859  ·            1  ·          -  │
·····································|·························|·············|·············|·············|···············|··············
|  ERC721AGasReporterMock            ·  transferTenAvg         ·          -  ·          -  ·     298706  ·            1  ·          -  │
·····································|·························|·············|·············|·············|···············|··············
|  ERC721AGasReporterMock            ·  transferTenDesc        ·          -  ·          -  ·     304523  ·            1  ·          -  │
·····································|·························|·············|·············|·············|···············|··············
|  ERC721AMock                       ·  approve                ·      50484  ·      52821  ·      51267  ·            3  ·          -  │
·····································|·························|·············|·············|·············|···············|··············
|  ERC721AMock                       ·  burn                   ·          -  ·          -  ·      60360  ·            1  ·          -  │
·····································|·························|·············|·············|·············|···············|··············
|  ERC721AMock                       ·  burn                   ·          -  ·          -  ·      60555  ·            1  ·          -  │
·····································|·························|·············|·············|·············|···············|··············
|  ERC721AMock                       ·  initializeOwnershipAt  ·      23706  ·      50575  ·      39827  ·            5  ·          -  │
·····································|·························|·············|·············|·············|···············|··············
|  ERC721AMock                       ·  mint                   ·      90563  ·      98443  ·      95455  ·           29  ·          -  │
·····································|·························|·············|·············|·············|···············|··············
|  ERC721AMock                       ·  safeMint               ·          -  ·          -  ·      99665  ·            2  ·          -  │
·····································|·························|·············|·············|·············|···············|··············
|  ERC721AMock                       ·  safeMint               ·      78156  ·     126107  ·      88467  ·          204  ·          -  │
·····································|·························|·············|·············|·············|···············|··············
|  ERC721AMock                       ·  safeTransferFrom       ·      86569  ·      91483  ·      89376  ·           14  ·          -  │
·····································|·························|·············|·············|·············|···············|··············
|  ERC721AMock                       ·  setApprovalForAll      ·      46229  ·      46241  ·      46240  ·           28  ·          -  │
·····································|·························|·············|·············|·············|···············|··············
|  ERC721AMock                       ·  setAux                 ·      27139  ·      44323  ·      32867  ·            3  ·          -  │
·····································|·························|·············|·············|·············|···············|··············
|  ERC721AMock                       ·  transferFrom           ·      45027  ·      83893  ·      79022  ·           15  ·          -  │
·····································|·························|·············|·············|·············|···············|··············
|  ERC721AQueryableMock              ·  burn                   ·      82798  ·     102106  ·      94383  ·            5  ·          -  │
·····································|·························|·············|·············|·············|···············|··············
|  ERC721AQueryableMock              ·  safeMint               ·      77635  ·      93301  ·      82307  ·           68  ·          -  │
·····································|·························|·············|·············|·············|···············|··············
|  ERC721AQueryableMock              ·  transferFrom           ·      83861  ·     103169  ·      95446  ·            5  ·          -  │
·····································|·························|·············|·············|·············|···············|··············
|  ERC721AQueryableStartTokenIdMock  ·  burn                   ·      84935  ·     104243  ·      96520  ·            5  ·          -  │
·····································|·························|·············|·············|·············|···············|··············
|  ERC721AQueryableStartTokenIdMock  ·  safeMint               ·      76201  ·      80135  ·      78032  ·           68  ·          -  │
·····································|·························|·············|·············|·············|···············|··············
|  ERC721AQueryableStartTokenIdMock  ·  transferFrom           ·      85998  ·     105306  ·      97583  ·            5  ·          -  │
·····································|·························|·············|·············|·············|···············|··············
|  ERC721AStartTokenIdMock           ·  approve                ·      52700  ·      55037  ·      53483  ·            3  ·          -  │
·····································|·························|·············|·············|·············|···············|··············
|  ERC721AStartTokenIdMock           ·  burn                   ·          -  ·          -  ·      62487  ·            1  ·          -  │
·····································|·························|·············|·············|·············|···············|··············
|  ERC721AStartTokenIdMock           ·  burn                   ·          -  ·          -  ·      62682  ·            1  ·          -  │
·····································|·························|·············|·············|·············|···············|··············
|  ERC721AStartTokenIdMock           ·  initializeOwnershipAt  ·      23684  ·      52690  ·      41088  ·            5  ·          -  │
·····································|·························|·············|·············|·············|···············|··············
|  ERC721AStartTokenIdMock           ·  mint                   ·      73453  ·      81321  ·      78337  ·           29  ·          -  │
·····································|·························|·············|·············|·············|···············|··············
|  ERC721AStartTokenIdMock           ·  safeMint               ·          -  ·          -  ·      82543  ·            2  ·          -  │
·····································|·························|·············|·············|·············|···············|··············
|  ERC721AStartTokenIdMock           ·  safeMint               ·      76179  ·     108985  ·      80733  ·          204  ·          -  │
·····································|·························|·············|·············|·············|···············|··············
|  ERC721AStartTokenIdMock           ·  safeTransferFrom       ·      88772  ·      93686  ·      91580  ·           14  ·          -  │
·····································|·························|·············|·············|·············|···············|··············
|  ERC721AStartTokenIdMock           ·  setApprovalForAll      ·      46207  ·      46219  ·      46219  ·           28  ·          -  │
·····································|·························|·············|·············|·············|···············|··············
|  ERC721AStartTokenIdMock           ·  setAux                 ·      27117  ·      44301  ·      32845  ·            3  ·          -  │
·····································|·························|·············|·············|·············|···············|··············
|  ERC721AStartTokenIdMock           ·  transferFrom           ·      47154  ·      86008  ·      81138  ·           15  ·          -  │
·····································|·························|·············|·············|·············|···············|··············
|  Deployments                                                 ·                                         ·  % of limit   ·             │
·······························································|·············|·············|·············|···············|··············
|  ERC721ABurnableMock                                         ·          -  ·          -  ·    1122636  ·        3.7 %  ·          -  │
·······························································|·············|·············|·············|···············|··············
|  ERC721ABurnableStartTokenIdMock                             ·          -  ·          -  ·    1179395  ·        3.9 %  ·          -  │
·······························································|·············|·············|·············|···············|··············
|  ERC721AGasReporterMock                                      ·          -  ·          -  ·    1065715  ·        3.6 %  ·          -  │
·······························································|·············|·············|·············|···············|··············
|  ERC721AMock                                                 ·          -  ·          -  ·    1308563  ·        4.4 %  ·          -  │
·······························································|·············|·············|·············|···············|··············
|  ERC721AQueryableMock                                        ·          -  ·          -  ·    1440762  ·        4.8 %  ·          -  │
·······························································|·············|·············|·············|···············|··············
|  ERC721AQueryableStartTokenIdMock                            ·          -  ·          -  ·    1505953  ·          5 %  ·          -  │
·······························································|·············|·············|·············|···············|··············
|  ERC721AStartTokenIdMock                                     ·          -  ·          -  ·    1369438  ·        4.6 %  ·          -  │
·······························································|·············|·············|·············|···············|··············
|  ERC721ReceiverMock                                          ·     296061  ·     296085  ·     296084  ·          1 %  ·          -  │
·--------------------------------------------------------------|-------------|-------------|-------------|---------------|-------------·
</pre>
</details>


<details><summary>This PR</summary>
<pre>
·--------------------------------------------------------------|---------------------------|-------------|-----------------------------·
|                     Solc version: 0.8.11                     ·  Optimizer enabled: true  ·  Runs: 800  ·  Block limit: 30000000 gas  │
·······························································|···························|·············|······························
|  Methods                                                                                                                             │
·····································|·························|·············|·············|·············|···············|··············
|  Contract                          ·  Method                 ·  Min        ·  Max        ·  Avg        ·  # calls      ·  usd (avg)  │
·····································|·························|·············|·············|·············|···············|··············
|  ERC721ABurnableMock               ·  approve                ·      30390  ·      50530  ·      40460  ·            2  ·          -  │
·····································|·························|·············|·············|·············|···············|··············
|  ERC721ABurnableMock               ·  burn                   ·      43329  ·     110975  ·      89981  ·           32  ·          -  │
·····································|·························|·············|·············|·············|···············|··············
|  ERC721ABurnableMock               ·  safeMint               ·          -  ·          -  ·     110982  ·           18  ·          -  │
·····································|·························|·············|·············|·············|···············|··············
|  ERC721ABurnableMock               ·  setApprovalForAll      ·      26277  ·      46189  ·      36233  ·            2  ·          -  │
·····································|·························|·············|·············|·············|···············|··············
|  ERC721ABurnableMock               ·  transferFrom           ·      83849  ·      89596  ·      86723  ·            2  ·          -  │
·····································|·························|·············|·············|·············|···············|··············
|  ERC721ABurnableStartTokenIdMock   ·  approve                ·      32527  ·      52667  ·      42597  ·            2  ·          -  │
·····································|·························|·············|·············|·············|···············|··············
|  ERC721ABurnableStartTokenIdMock   ·  burn                   ·      45444  ·     110879  ·      89565  ·           34  ·          -  │
·····································|·························|·············|·············|·············|···············|··············
|  ERC721ABurnableStartTokenIdMock   ·  safeMint               ·          -  ·          -  ·      93860  ·           18  ·          -  │
·····································|·························|·············|·············|·············|···············|··············
|  ERC721ABurnableStartTokenIdMock   ·  setApprovalForAll      ·      26255  ·      46167  ·      36211  ·            2  ·          -  │
·····································|·························|·············|·············|·············|···············|··············
|  ERC721ABurnableStartTokenIdMock   ·  transferFrom           ·      85986  ·      89522  ·      87754  ·            2  ·          -  │
·····································|·························|·············|·············|·············|···············|··············
|  ERC721AGasReporterMock            ·  mintOne                ·      56188  ·      90388  ·      59608  ·           10  ·          -  │
·····································|·························|·············|·············|·············|···············|··············
|  ERC721AGasReporterMock            ·  mintTen                ·      73869  ·     108069  ·      97809  ·           10  ·          -  │
·····································|·························|·············|·············|·············|···············|··············
|  ERC721AGasReporterMock            ·  safeMintOne            ·      58938  ·      93138  ·      76038  ·            2  ·          -  │
·····································|·························|·············|·············|·············|···············|··············
|  ERC721AGasReporterMock            ·  safeMintTen            ·      76620  ·     110820  ·      93720  ·            2  ·          -  │
·····································|·························|·············|·············|·············|···············|··············
|  ERC721AGasReporterMock            ·  transferFrom           ·      44392  ·      86094  ·      54818  ·            4  ·          -  │
·····································|·························|·············|·············|·············|···············|··············
|  ERC721AGasReporterMock            ·  transferTenAsc         ·          -  ·          -  ·     297859  ·            1  ·          -  │
·····································|·························|·············|·············|·············|···············|··············
|  ERC721AGasReporterMock            ·  transferTenAvg         ·          -  ·          -  ·     298736  ·            1  ·          -  │
·····································|·························|·············|·············|·············|···············|··············
|  ERC721AGasReporterMock            ·  transferTenDesc        ·          -  ·          -  ·     304658  ·            1  ·          -  │
·····································|·························|·············|·············|·············|···············|··············
|  ERC721AMock                       ·  approve                ·      50484  ·      52821  ·      51267  ·            3  ·          -  │
·····································|·························|·············|·············|·············|···············|··············
|  ERC721AMock                       ·  burn                   ·          -  ·          -  ·      60360  ·            1  ·          -  │
·····································|·························|·············|·············|·············|···············|··············
|  ERC721AMock                       ·  burn                   ·          -  ·          -  ·      60555  ·            1  ·          -  │
·····································|·························|·············|·············|·············|···············|··············
|  ERC721AMock                       ·  initializeOwnershipAt  ·      23706  ·      50581  ·      39831  ·            5  ·          -  │
·····································|·························|·············|·············|·············|···············|··············
|  ERC721AMock                       ·  mint                   ·      90563  ·      98443  ·      95455  ·           29  ·          -  │
·····································|·························|·············|·············|·············|···············|··············
|  ERC721AMock                       ·  safeMint               ·          -  ·          -  ·      99665  ·            2  ·          -  │
·····································|·························|·············|·············|·············|···············|··············
|  ERC721AMock                       ·  safeMint               ·      78156  ·     126107  ·      88467  ·          204  ·          -  │
·····································|·························|·············|·············|·············|···············|··············
|  ERC721AMock                       ·  safeTransferFrom       ·      86569  ·      91483  ·      89376  ·           14  ·          -  │
·····································|·························|·············|·············|·············|···············|··············
|  ERC721AMock                       ·  setApprovalForAll      ·      46229  ·      46241  ·      46240  ·           28  ·          -  │
·····································|·························|·············|·············|·············|···············|··············
|  ERC721AMock                       ·  setAux                 ·      27139  ·      44323  ·      32867  ·            3  ·          -  │
·····································|·························|·············|·············|·············|···············|··············
|  ERC721AMock                       ·  transferFrom           ·      45027  ·      83893  ·      79022  ·           15  ·          -  │
·····································|·························|·············|·············|·············|···············|··············
|  ERC721AQueryableMock              ·  burn                   ·      82798  ·     102109  ·      94385  ·            5  ·          -  │
·····································|·························|·············|·············|·············|···············|··············
|  ERC721AQueryableMock              ·  safeMint               ·      77635  ·      93301  ·      82307  ·           68  ·          -  │
·····································|·························|·············|·············|·············|···············|··············
|  ERC721AQueryableMock              ·  transferFrom           ·      83861  ·     103172  ·      95448  ·            5  ·          -  │
·····································|·························|·············|·············|·············|···············|··············
|  ERC721AQueryableStartTokenIdMock  ·  burn                   ·      84935  ·     104246  ·      96522  ·            5  ·          -  │
·····································|·························|·············|·············|·············|···············|··············
|  ERC721AQueryableStartTokenIdMock  ·  safeMint               ·      76201  ·      80135  ·      78032  ·           68  ·          -  │
·····································|·························|·············|·············|·············|···············|··············
|  ERC721AQueryableStartTokenIdMock  ·  transferFrom           ·      85998  ·     105309  ·      97585  ·            5  ·          -  │
·····································|·························|·············|·············|·············|···············|··············
|  ERC721AStartTokenIdMock           ·  approve                ·      52700  ·      55037  ·      53483  ·            3  ·          -  │
·····································|·························|·············|·············|·············|···············|··············
|  ERC721AStartTokenIdMock           ·  burn                   ·          -  ·          -  ·      62487  ·            1  ·          -  │
·····································|·························|·············|·············|·············|···············|··············
|  ERC721AStartTokenIdMock           ·  burn                   ·          -  ·          -  ·      62682  ·            1  ·          -  │
·····································|·························|·············|·············|·············|···············|··············
|  ERC721AStartTokenIdMock           ·  initializeOwnershipAt  ·      23684  ·      52696  ·      41091  ·            5  ·          -  │
·····································|·························|·············|·············|·············|···············|··············
|  ERC721AStartTokenIdMock           ·  mint                   ·      73453  ·      81321  ·      78337  ·           29  ·          -  │
·····································|·························|·············|·············|·············|···············|··············
|  ERC721AStartTokenIdMock           ·  safeMint               ·          -  ·          -  ·      82543  ·            2  ·          -  │
·····································|·························|·············|·············|·············|···············|··············
|  ERC721AStartTokenIdMock           ·  safeMint               ·      76179  ·     108985  ·      80733  ·          204  ·          -  │
·····································|·························|·············|·············|·············|···············|··············
|  ERC721AStartTokenIdMock           ·  safeTransferFrom       ·      88772  ·      93686  ·      91580  ·           14  ·          -  │
·····································|·························|·············|·············|·············|···············|··············
|  ERC721AStartTokenIdMock           ·  setApprovalForAll      ·      46207  ·      46219  ·      46219  ·           28  ·          -  │
·····································|·························|·············|·············|·············|···············|··············
|  ERC721AStartTokenIdMock           ·  setAux                 ·      27117  ·      44301  ·      32845  ·            3  ·          -  │
·····································|·························|·············|·············|·············|···············|··············
|  ERC721AStartTokenIdMock           ·  transferFrom           ·      47154  ·      86008  ·      81138  ·           15  ·          -  │
·····································|·························|·············|·············|·············|···············|··············
|  Deployments                                                 ·                                         ·  % of limit   ·             │
·······························································|·············|·············|·············|···············|··············
|  ERC721ABurnableMock                                         ·          -  ·          -  ·    1124376  ·        3.7 %  ·          -  │
·······························································|·············|·············|·············|···············|··············
|  ERC721ABurnableStartTokenIdMock                             ·          -  ·          -  ·    1181327  ·        3.9 %  ·          -  │
·······························································|·············|·············|·············|···············|··············
|  ERC721AGasReporterMock                                      ·          -  ·          -  ·    1067467  ·        3.6 %  ·          -  │
·······························································|·············|·············|·············|···············|··············
|  ERC721AMock                                                 ·          -  ·          -  ·    1310315  ·        4.4 %  ·          -  │
·······························································|·············|·············|·············|···············|··············
|  ERC721AQueryableMock                                        ·          -  ·          -  ·    1442514  ·        4.8 %  ·          -  │
·······························································|·············|·············|·············|···············|··············
|  ERC721AQueryableStartTokenIdMock                            ·          -  ·          -  ·    1507885  ·          5 %  ·          -  │
·······························································|·············|·············|·············|···············|··············
|  ERC721AStartTokenIdMock                                     ·          -  ·          -  ·    1371389  ·        4.6 %  ·          -  │
·······························································|·············|·············|·············|···············|··············
|  ERC721ReceiverMock                                          ·     296061  ·     296085  ·     296084  ·          1 %  ·          -  │
·--------------------------------------------------------------|-------------|-------------|-------------|---------------|-------------·
</pre>
</details>

**Comment by kadenzipfel** [comment]
yep, looks like i made some false assumptions. will keep digging

**Pull request closed without merging**